pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisStreamMetadataProvider.java

```diff
@@ -0,0 +1,189 @@
+      if (currentEndOffset != null) { // Segment DONE (committing/committed)
+        String endingSequenceNumber = shard.sequenceNumberRange().endingSequenceNumber();
+        if (endingSequenceNumber != null) { // Shard has ended, check if we're also done consuming it
+          if (consumedEndOfShard(currentEndOffset, currentPartitionGroupConsumptionStatus)) {
+            shardsEnded.add(shardId);
+            continue; // Shard ended and we're done consuming it. Skip
+          }
+        }
+        newStartOffset = currentEndOffset;
+      } else { // Segment IN_PROGRESS
+        newStartOffset = currentPartitionGroupConsumptionStatus.getStartOffset();
+      }
+      newPartitionGroupMetadataList.add(
+          new PartitionGroupMetadata(currentPartitionGroupConsumptionStatus.getPartitionGroupId(), newStartOffset));
+    }
+
+    // Add brand new shards
+    for (Map.Entry<String, Shard> entry : shardIdToShardMap.entrySet()) {
+      // If shard was already in current list, skip
+      String newShardId = entry.getKey();
+      if (shardsInCurrent.contains(newShardId)) {
+        continue;
+      }
+      StreamPartitionMsgOffset newStartOffset;
+      Shard newShard = entry.getValue();
+      String parentShardId = newShard.parentShardId();
+
+      // Add the new shard in the following 3 cases:
+      // 1. Root shards - Parent shardId will be null. Will find this case when creating new table.
+      // 2. Parent expired - Parent shardId will not be part of shardIdToShard map
+      // 3. Parent reached EOL and completely consumed.
+      if (parentShardId == null || !shardIdToShardMap.containsKey(parentShardId) || shardsEnded.contains(parentShardId)) {
+        Map<String, String> shardToSequenceNumberMap = new HashMap<>();
+        shardToSequenceNumberMap.put(newShardId, newShard.sequenceNumberRange().startingSequenceNumber());
+        newStartOffset = new KinesisPartitionGroupOffset(shardToSequenceNumberMap);
+        int partitionGroupId = getPartitionGroupIdFromShardId(newShardId);
+        newPartitionGroupMetadataList.add(new PartitionGroupMetadata(partitionGroupId, newStartOffset));
+      }
+    }
+    return newPartitionGroupMetadataList;
+  }
+
+  /**
+   * Converts a shardId string to a partitionGroupId integer by parsing the digits of the shardId
+   * e.g. "shardId-000000000001" becomes 1
+   * FIXME: Although practically the shard values follow this format, the Kinesis docs don't guarantee it.
+   *  Re-evaluate if this convention needs to be changed.
+   */
+  private int getPartitionGroupIdFromShardId(String shardId) {
+    String shardIdNum = StringUtils.stripStart(StringUtils.removeStart(shardId, SHARD_ID_PREFIX), "0");
+    return shardIdNum.isEmpty() ? 0 : Integer.parseInt(shardIdNum);
+  }
+
+  private boolean consumedEndOfShard(StreamPartitionMsgOffset startCheckpoint, PartitionGroupConsumptionStatus partitionGroupConsumptionStatus)
```

Comment:
Can you apply the formatting for this class? I believe that this line probably runs over the line charact limit.